Customer Segmentation Strategy

The customer base is segmented using the RFM framework.

Segments include:

- Champions
- Loyal Customers
- Potential Loyalists
- New Customers
- Promising
- Need Attention
- At Risk
- Can't Lose Them
- Hibernating
- Lost

The objective is to enable targeted marketing strategies based on customer behavior and value.

In [1]:
import pandas as pd

In [2]:
customer_rfm = pd.read_csv('../data/processed/customer_rfm.csv')

In [3]:
customer_rfm['Segment'] = 'Others'

In [4]:
champion_mask = (
    (customer_rfm['R_Score'] >= 4) &
    (customer_rfm['F_Score'] >= 4) &
    (customer_rfm['M_Score'] >= 4)
)


customer_rfm.loc[champion_mask, 'Segment'] = 'Champions'

In [5]:
customer_rfm['Segment'].value_counts()

Segment
Others       4144
Champions     856
Name: count, dtype: int64

Champions 

Objective

Identify the most valuable customers who have purchased recently, buy frequently, and generate the highest revenue.

Finding

A total of 856 customers were classified as Champions.

These customers achieved high scores across all three RFM dimensions:

High Recency → They purchased recently.
High Frequency → They purchase frequently.
High Monetary Value → They spend significantly more than other customers.
Business Insight

Champions represent the company's most valuable customer segment. They contribute substantially to overall revenue while also demonstrating strong engagement and purchase loyalty. Retaining these customers is generally more cost-effective than acquiring new ones.

Business Recommendation
- Provide VIP or premium membership benefits.
- Offer early access to new products and exclusive promotions.
- Deliver personalized recommendations and premium customer support.
- Use this segment for referral and ambassador programs to attract new customers.

In [6]:
loyal_mask =(
    (customer_rfm['Segment'] == 'Others') &
    (customer_rfm['F_Score'] >= 4) &
    (customer_rfm['R_Score'] >= 3)
)

customer_rfm.loc[loyal_mask, 'Segment'] = 'Loyal Customer'

In [7]:
customer_rfm['Segment'].value_counts()

Segment
Others            3394
Champions          856
Loyal Customer     750
Name: count, dtype: int64

Loyal Customers

Objective

Identify customers who purchase frequently and remain actively engaged with the business.

Finding

751 customers were classified as Loyal Customers.

These customers exhibit high purchase frequency and relatively recent activity, indicating consistent engagement with the platform.

Business Insight

Loyal Customers represent a stable source of recurring revenue. While they may not always be the highest spenders, their consistent purchasing behavior makes them valuable for long-term customer retention.

Business Recommendation

- Reward loyalty through membership programs.
- Provide exclusive discounts and personalized offers.
- Encourage progression toward the Champions segment through targeted campaigns.

In [8]:
potential_loyalists_mask =(
    (customer_rfm['Segment'] == 'Others') &
    (customer_rfm['F_Score'].isin([2,3])) &
    (customer_rfm['R_Score'] >= 4)
)

customer_rfm.loc[potential_loyalists_mask, 'Segment'] = 'Potential Loyalists'

In [9]:
customer_rfm['Segment'].value_counts()

Segment
Others                 2742
Champions               856
Loyal Customer          750
Potential Loyalists     652
Name: count, dtype: int64

Potential Loyalists

Objective

Identify recently acquired customers who have started engaging with the business and show strong potential to become loyal customers.

Finding

A total of 652 customers were classified as Potential Loyalists.

These customers have made recent purchases but have not yet reached a high purchase frequency, indicating that they are still in the early stages of their customer journey.

Business Insight

Potential Loyalists represent a valuable growth opportunity. With appropriate engagement strategies, many of these customers can transition into the Loyal Customer or Champion segments, increasing their lifetime value.

Business Recommendation
- Send personalized follow-up emails after purchases.
- Offer limited-time discounts to encourage repeat purchases.
- Recommend related products based on previous orders.
- Enroll customers in loyalty or rewards programs early to strengthen long-term engagement.

In [10]:
new_customer_mask =(
    (customer_rfm['Segment'] == 'Others') &
    (customer_rfm['F_Score'] == 1) &
    (customer_rfm['R_Score'] >= 4)
)

customer_rfm.loc[new_customer_mask, 'Segment'] = 'New Customer'

In [11]:
customer_rfm['Segment'].value_counts()

Segment
Others                 2560
Champions               856
Loyal Customer          750
Potential Loyalists     652
New Customer            182
Name: count, dtype: int64

New Customers

Objective

Identify customers who have recently made their first purchase and have just entered the customer lifecycle.

Finding

A total of 182 customers were identified as New Customers.

These customers have purchased very recently but have not yet established a repeat purchasing pattern.

Business Insight

New Customers represent the first stage of the customer journey. Their future value depends on whether the business can successfully encourage additional purchases and build long-term engagement.

Business Recommendation

- Send a welcome email after the first purchase.
- Offer a discount for the second purchase.
- Recommend complementary products.
- Provide an excellent onboarding and post-purchase experience.

In [12]:
cant_lose_mask = (
    (customer_rfm['Segment'] == 'Others') &
    (customer_rfm['R_Score'] <= 2) &
    (customer_rfm['F_Score'] >= 4) &
    (customer_rfm['M_Score'] >= 4)
)

customer_rfm.loc[ cant_lose_mask, 'Segment'] = 'Cannot Lose Them'

In [13]:
customer_rfm['Segment'].value_counts()

Segment
Others                 2315
Champions               856
Loyal Customer          750
Potential Loyalists     652
Cannot Lose Them        245
New Customer            182
Name: count, dtype: int64

In [14]:
at_risk_mask =(
    (customer_rfm['Segment'] == 'Others') &
    (customer_rfm['F_Score'] >= 3) &
    (customer_rfm['R_Score'] <= 2)
)

customer_rfm.loc[at_risk_mask, 'Segment'] = 'At Risk'

In [15]:
customer_rfm['Segment'].value_counts()

Segment
Others                 1782
Champions               856
Loyal Customer          750
Potential Loyalists     652
At Risk                 533
Cannot Lose Them        245
New Customer            182
Name: count, dtype: int64

⚠️ At Risk

Objective

Identify customers who were previously active purchasers but have not made a recent purchase and are at risk of churning.

Finding

A total of 778 customers were classified as At Risk.

These customers have a history of frequent purchases but have shown low recent activity, indicating a high probability of customer churn if no action is taken.

Business Insight

At Risk customers have already demonstrated strong engagement in the past, making them more valuable than newly acquired customers. Re-engaging these customers is often more cost-effective than acquiring new ones.

Business Recommendation

- Launch win-back campaigns with personalized offers.
- Send reminder emails highlighting new products or promotions.
- Provide limited-time discounts to encourage reactivation.
- Prioritize high-value At Risk customers for personalized outreach.

In [16]:
hibernating_mask =(
    (customer_rfm['Segment'] == 'Others') &
    (customer_rfm['F_Score'] <= 2) &
    (customer_rfm['R_Score'] == 2)
)

customer_rfm.loc[hibernating_mask, 'Segment'] = 'Hibernating'

In [17]:
customer_rfm['Segment'].value_counts()

Segment
Others                 1317
Champions               856
Loyal Customer          750
Potential Loyalists     652
At Risk                 533
Hibernating             465
Cannot Lose Them        245
New Customer            182
Name: count, dtype: int64

In [18]:
customer_rfm[
    (customer_rfm['R_Score'] <= 2) &
    (customer_rfm['F_Score'] <= 2)
].shape

(1218, 11)

In [19]:
customer_rfm.groupby('Segment')[['Recency', 'Frequency', 'Monetary']].mean().round(1)

,Recency,Frequency,Monetary
Segment,,,
At Risk,205.2,3.2,2993.1
Cannot Lose Them,175.6,4.9,8832.9
Champions,30.1,6.4,9776.5
Hibernating,167.2,1.6,2294.5
Loyal Customer,65.7,5.1,4567.9
New Customer,36.2,1.2,1341.6
Others,223.1,1.7,2086.7
Potential Loyalists,35.4,2.5,3321.7


In [20]:
lost_mask = (
    (customer_rfm['Segment'] == 'Others') &
    (customer_rfm['R_Score'] == 1) &
    (customer_rfm['F_Score'] <= 2)
)

customer_rfm.loc[lost_mask, 'Segment'] = 'Lost'

In [21]:
customer_rfm['Segment'].value_counts()

Segment
Champions              856
Lost                   753
Loyal Customer         750
Potential Loyalists    652
Others                 564
At Risk                533
Hibernating            465
Cannot Lose Them       245
New Customer           182
Name: count, dtype: int64

In [22]:
customer_rfm['R_Score'].value_counts().sort_index()

R_Score
1     999
2     997
3     972
4    1027
5    1005
Name: count, dtype: int64

In [23]:
pd.crosstab(
    customer_rfm['R_Score'],
    customer_rfm['Segment']
)

Segment,At Risk,Cannot Lose Them,Champions,Hibernating,Lost,Loyal Customer,New Customer,Others,Potential Loyalists
R_Score,,,,,,,,,
1,207,39,0,0,753,0,0,0,0
2,326,206,0,465,0,0,0,0,0
3,0,0,0,0,0,408,0,564,0
4,0,0,394,0,0,164,104,0,365
5,0,0,462,0,0,178,78,0,287


### 📌 Customer Segmentation Summary

In this section, customers were segmented using the **RFM (Recency, Frequency, Monetary)** framework. Rather than relying solely on predefined templates, the segmentation rules were refined to better reflect the customer lifecycle and business objectives.

The final customer segments are:

* **Champions:** Recent, frequent, and high-value customers. They represent the company's most valuable customer group.
* **Loyal Customers:** Customers who purchase regularly and maintain a strong relationship with the business.
* **Potential Loyalists:** Customers who have shown promising purchasing behavior and have the potential to become loyal customers with appropriate engagement.
* **New Customers:** Recently acquired customers with limited purchase history who require nurturing.
* **Cannot Lose Them:** High-value customers who have become inactive and require immediate retention efforts.
* **At Risk:** Customers who previously engaged with the business but have recently reduced their activity and may churn.
* **Hibernating:** Customers with low recent activity who still have potential to be reactivated through targeted marketing campaigns.
* **Lost:** Customers who have not purchased for a long period and exhibit very low engagement, indicating a low probability of returning.

This segmentation provides a solid foundation for customer analysis and enables the development of personalized marketing strategies, customer retention campaigns, and business decision-making based on customer value and behavior.


In [24]:
customer_rfm.to_csv('../data/processed/customer_final.csv', index=False)